### PR TITLE
Add initial draft of GTScript parallel model

### PR DESCRIPTION
**Pull request by egparedes**
This is a initial draft to discuss and formalize the parallel model we want to implement in GT4Py/GTScript, based on the Dawn parallel model (https://github.com/MeteoSwiss-APN/dawn/wiki/SIR-Execution-Model). The current implementation does not fully verify this model.

### CODE REVIEW

File: GTScript-Parallel-model.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+Every statement inside a vertical interval is executed on the full horizontal `IJ` plane before the next one is executed. Individual statements of a computation are considered embarrassingly parallel in the horizontal and therefore there are no guarantees about the loop order in the `IJ` plane. Every iteration of the `K`-loop defined by a vertical interval can only start after the previous iteration have finished. Vertical intervals defined in a `computation` will be sorted according to the iteration policy and then executed in order, that is, every vertical interval will start its execution after all the previous vertical intervals in the iteration order have been fully executed. In the case of a `PARALLEL` iteration policy, vertical intervals will be ordered in any implementation-specific way (no guarantees).
+
+To summarize the model, we can say that:
+
+- *computations* are executed sequentially in the order they appear in the code,
+- vertical *intervals* are executed sequentially in the order defined by the *iteration policy* of the  *computation* (where `PARALLEL` means that order is not relevant and thus it will be implementation-specific),
```
Comment (gronerl) [inline GTScript-Parallel-model.md:10]: Is it enough to say that intervals in a `PARALLEL` computation can be reordered to allow us to actually schedule them independently? At least the "sequentially" on line 10 I would remove, after that I think it would be equivalent to not making any assumption on the order.

File: GTScript-Parallel-model.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+- *statements* inside *intervals* are executed as (sequential) for-loops
```
Comment (gronerl) [inline GTScript-Parallel-model.md:11]: not sequential if PARALLEL.

File: GTScript-Parallel-model.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+A GTScript `computation` is composed by one or more vertical `interval` specifications, each one of them representing a vertical loop over its vertical iteration range with a given iteration policy (`FORWARD`, `BACKWARD`, `PARALLEL`). Vertical `interval`s defined inside a computation should not have overlapping iteration ranges.
+
+Every statement inside a vertical interval is executed on the full horizontal `IJ` plane before the next one is executed. Individual statements of a computation are considered embarrassingly parallel in the horizontal and therefore there are no guarantees about the loop order in the `IJ` plane. Every iteration of the `K`-loop defined by a vertical interval can only start after the previous iteration have finished. Vertical intervals defined in a `computation` will be sorted according to the iteration policy and then executed in order, that is, every vertical interval will start its execution after all the previous vertical intervals in the iteration order have been fully executed. In the case of a `PARALLEL` iteration policy, vertical intervals will be ordered in any implementation-specific way (no guarantees).
```
Comment (gronerl) [inline GTScript-Parallel-model.md:5]:
```suggestion
The generated code will be equivalent to executing every statement inside a vertical interval on the full horizontal `IJ` plane before the next one is executed. Individual statements of a computation are considered embarrassingly parallel in the horizontal and therefore there are no guarantees about the loop order in the `IJ` plane. Every iteration of the `K`-loop defined by a vertical interval can only start after the previous iteration have finished. Vertical intervals defined in a `computation` will be sorted according to the iteration policy and then executed in order, that is, every vertical interval will start its execution after all the previous vertical intervals in the iteration order have been fully executed. In the case of a `PARALLEL` iteration policy, vertical intervals will be ordered in any implementation-specific way (no guarantees).
```

Comment (gronerl) [reply GTScript-Parallel-model.md:5]: Changed only the start of the line to indicate that we are free to optimize this as long s it's equivalent

File: GTScript-Parallel-model.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+parfor k in range(start, end):
+    parfor ij:
+        a[i, j, k] = tmp[i+1, j+1, k]
+    parfor ij:
+        b[i, j, k] = 2 * a[i, j, k]
+```
+
+where `parfor` means that there is no guarantee of the order in which the iteration is performed. Additionally, the following restrictions apply:
+
+- No self-assignments if `i != 0` or `j != 0`
```
Comment (gronerl) [inline GTScript-Parallel-model.md:70]: Can be expanded into 
* what self-assignment means (namely, reading and writing withtin the same interval, not just the same statement)
* `k!=0` is also not allowed if PARALLEL
* and that the `i!=0` is understood in the read offsets

Comment (havogt) [reply GTScript-Parallel-model.md:70]: I think actually the whole sentence 

> Additionally, the following restrictions apply: ...

should be removed as it doesn't add extra information. Or if you like to have this information somewhere, I'd rephrase this as a consequence from the parallel model.

File: GTScript-Parallel-model.md (hunk 1 above)
Comment (havogt) [reply GTScript-Parallel-model.md:10]: Probably we should discuss the future of `PARALLEL` as it doesn't add extra semantic information to the parallel model, unless we will (also in the future) blindly apply `PARALLEL` instead of checking.

Currently `PARALLEL` doesn't imply anything: it doesn't have to be executed in parallel, it can be executed forward or backward as well.

Comment (egparedes) [reply GTScript-Parallel-model.md:10]: Just a comment: the future of PARALLEL is one of the issues I had collected for the workshop: https://hackmd.io/@egparedes/Bkf7qxibv

File: GTScript-Parallel-model.md (hunk 4 above)
Comment (egparedes) [reply GTScript-Parallel-model.md:70]: done

**Comment by egparedes** [comment]
@gronerl  @havogt I updated the document following your suggestions. I propose to merge this version now so we have it in `master` and try to remove or fix the references to `PARALLEL` from there.
